Total Rework to add the ability for expansion of the program.

phantomstranger opened the pull request:
I tried testing the program, but I am stuck and red about it. I would love to see more features added. Maybe adding the log file to a webpage that gets refreshed to show past events of the sensor log. I would love to have a RPi security system. Maybe even add output to a speaker for audible response to sensors or events. Maybe even an external input from a keypad to arm/disarm. I am VERY new to programming in general. I know just enough to be dangerous. LOL The files are supposed to be pipins.conf, email.conf and log.txt. It wouldn't let me upload a ".conf" extension so I added the ".txt"
[email.conf.txt](https://github.com/Banrai/OpenSesamePi/files/1402725/email.conf.txt)
[pipins.conf.txt](https://github.com/Banrai/OpenSesamePi/files/1402726/pipins.conf.txt)

dpapathanasiou commented:
Thank you for the pull request.

I want to encourage you, in that it is great to see more contributions, but I cannot merge your version as-is.

First, here are some coding style suggestions:

WRT settings, such as sleep time, etc., it is better that those are defined outside the core logic, and imported separately.

That way, they can be overridden/adjusted with local settings outside of source control as needed.

Second, if your settings depend on loading data from text files, you should have a single file load method, and just pass the file paths for each setting as needed.

All those initial file loads should happen in an init() or main() method, so they're present at the start, and not unnecessarily reloaded each time a particular function is called.

Finally, I'd like to understand what else you'd like the script to do.

Collecting open/close events in a log is simple enough, and making data available through a web interface is also possible (see the PiScan project as an example), but I'm not sure what the overall goal is.

Could you articulate it a little more?

phantomstranger commented:
I would eventually like to add features for specific sensors (pin as input) would trip another pin (as output) to turn on a light, siren, or ect. Alternatively make multiple input events control one ouput. Basically I would like to eventually like to have a "drop-in" security system. I could use a "softphone" or other device to dialout to security monitoring company like NextAlarm. I would like to have a opensource security system. It would make hardwire security systems cheaper. Maybe eventually add wireless devices but that is much later. I hope I haven't made the murky water into mud. This is a huge task but I feel it could be worthwhile and I'm sure with enough "advertising" that could bring in more developers to add more to it.

phantomstranger commented:
Is this any better?

`#!/usr/bin/env python
from gpiozero import Button
from signal import pause
from datetime import datetime
import subprocess
from string import Template

emails = open('email.conf', 'r')
pipins = open('pipins.conf', 'r')

notifylist = []
for word in emails.read().split():
    notifylist.append(word)

actpins = {}
for line in pipins:
    l_split = line.split()
    actpins[int(l_split[0])] = l_split[1]


def current_date(fmt="%a %d-%m-%Y @ %H:%M:%S"):
    return datetime.strftime(datetime.now(), fmt)


notifycmd = {pin: Template("""echo "$date $sensor $state" | mail -s "Pi: $sensor $state" $email""") for pin in ACTPINS}


def notify(id, state, sensor_name):
    """Send each of the email addresses in NOTIFY_LIST a message"""
    for email in notifylist:
        shell_cmd = notifycmd[id].substitute(date=current_date(), state=state, sensor=sensor_name, email=email)
        proc = subprocess.Popen(shell_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        stdout_value, stderr_value = proc.communicate()
        # This is where the sensor state changes are written to a file called sensor.log
        with open('sensor.log', 'a') as f:
            f.write('{}\n'.format(shell_cmd))


def dostate(button):
    # function to send email etc will only run when opened or closed
    state = 'closed' if button.is_pressed else 'opened'
    notify(button.pin, state, actpins[button.pin])
    print(state)


buttons = {}
for id in actpins:
    buttons[id] = Button(id)
    buttons[id].when_pressed = dostate
    buttons[id].when_released = dostate
    print(id)
    print(buttons)
    print(actpins)
    print(notifylist)

pause()
`

dpapathanasiou commented:
It's better, but nowhere near enough in terms of what's needed for the vision you've proposed.

Remember that the pi has physical limitations in terms of how many sensors and triggers are possible at the same time, b/c of the number of gpio and other interfaces.

So while it's possible to have light and sound sensors, alarms, cameras taking still photos and videos, while serving web pages, not all of the those things are simultaneously interoperable.

We should really take a step back and define which types of plugins are desired, and which combinations are feasible.
